File: src/Application/Services/Accountant/DebtService.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Application.Contracts.Accountant.Common.Models;
using Application.Contracts.Accountant.Debts;
using Application.Contracts.Accountant.Debts.Models;
using AutoMapper;
using Domain.Entities.Accountant;

namespace Application.Services.Accountant;

public class DebtService : IDebtService
{
    private readonly IDebtsRepository _debtsRepository;
    private readonly IMapper _mapper;

    public DebtService(
        IDebtsRepository debtsRepository,
        IMapper mapper)
    {
        _debtsRepository = debtsRepository;
        _mapper = mapper;
    }

    public IEnumerable<DebtDto> GetAll(GetAll model)
    {
        var debt = _debtsRepository.GetAll(model.UserId, model.FromModifiedDate);

        var debtDtos = debt.Select(x => _mapper.Map<DebtDto>(x));

        return debtDtos;
    }

    public IEnumerable<int> GetDeletedIds(GetDeletedIds model)
    {
        return _debtsRepository.GetDeletedIds(model.UserId, model.FromDate);
    }

    public Task<int> CreateAsync(CreateDebt model)
    {
        var debt = _mapper.Map<Debt>(model);

        debt.Person = debt.Person.Trim();
        if (debt.Description != null)
        {
            debt.Description = debt.Description.Trim();
        }

        return _debtsRepository.CreateAsync(debt);
    }

    public Task<int> CreateMergedAsync(CreateDebt model)
    {
        var debt = _mapper.Map<Debt>(model);

        debt.Person = debt.Person.Trim();
        if (debt.Description != null)
        {
            debt.Description = debt.Description.Trim();
        }

        return _debtsRepository.CreateMergedAsync(debt);
    }

    public async Task UpdateAsync(UpdateDebt model)
    {
        var debt = _mapper.Map<Debt>(model);

        debt.Person = debt.Person.Trim();
        if (debt.Description != null)
        {
            debt.Description = debt.Description.Trim();
        }

        await _debtsRepository.UpdateAsync(debt);
    }

    public async Task DeleteAsync(int id, int userId)
    {
        await _debtsRepository.DeleteAsync(id, userId);
    }
}

File: src/Application/Accountant.Application.Fs/Services/DebtService.fs
namespace Accountant.Application.Fs.Services

open Accountant.Domain.Models
open Accountant.Application.Fs
open Accountant.Application.Fs.Models.Debts

module DebtService =
    let mapAll (debts: seq<Debt>) : seq<DebtDto> =
        debts
        |> Seq.map (fun x ->
            { Id = x.Id
              Person = x.Person
              Amount = x.Amount
              Currency = x.Currency
              UserIsDebtor = x.UserIsDebtor
              Description = x.Description
              CreatedDate = x.CreatedDate
              ModifiedDate = x.ModifiedDate })

    let prepareForCreate (model: CreateDebt) (userId: int) =
        { Id = 0
          UserId = userId
          Person = model.Person.Trim()
          Amount = model.Amount
          Currency = model.Currency
          UserIsDebtor = model.UserIsDebtor
          Description = Utils.noneOrTrimmed model.Description
          CreatedDate = model.CreatedDate
          ModifiedDate = model.ModifiedDate }

    let prepareForCreateMerged (model: CreateDebt) (userId: int) =
        { Id = 0
          UserId = userId
          Person = model.Person.Trim()
          Amount = model.Amount
          Currency = model.Currency
          UserIsDebtor = model.UserIsDebtor
          Description = Utils.noneOrTrimmed model.Description
          CreatedDate = model.CreatedDate
          ModifiedDate = model.ModifiedDate }

    let prepareForUpdate (model: UpdateDebt) (userId: int) =
        { Id = model.Id
          UserId = userId
          Person = model.Person.Trim()
          Amount = model.Amount
          Currency = model.Currency
          UserIsDebtor = model.UserIsDebtor
          Description = Utils.noneOrTrimmed model.Description
          CreatedDate = model.CreatedDate
          ModifiedDate = model.ModifiedDate }
